packages/common/src/PredicateConnector.ts

```diff
@@ -0,0 +1,343 @@
+      const address = await this.getAccountAddress();
+      if (!address) {
+        continue;
+      }
+
+      const { fuelProvider } = await this.getProviders();
+      const predicate = predicateInstance.build(address, fuelProvider, [1]);
+
+      const balance = await predicate.getBalance();
+
+      if (balance.toString() !== bn(0).toString()) {
+        return predicateInstance;
+      }
+    }
+
+    return null;
+  }
+
+  protected getNewestPredicate(): Maybe<PredicateFactory> {
+    const predicateVersions = Object.entries(this.getPredicateVersions()).map(
+      ([key, pred]) => ({
+        pred,
+        key,
+      }),
+    );
+
+    const newest = predicateVersions.sort(
+      (a, b) => Number(b.pred.generatedAt) - Number(a.pred.generatedAt),
+    )[0];
+    if (!newest) return null;
+
+    return new PredicateFactory(
+      this.getWalletAdapter(),
+      newest.pred.predicate,
+      newest.key,
+    );
+  }
+
+  protected async setupPredicate(): Promise<PredicateFactory> {
+    if (this.customPredicate?.abi && this.customPredicate?.bin) {
+      this.predicateAccount = new PredicateFactory(
+        this.getWalletAdapter(),
+        this.customPredicate,
+        'custom',
+      );
+      this.predicateAddress = 'custom';
+
+      return this.predicateAccount;
+    }
+
+    const predicate =
+      (await this.getCurrentUserPredicate()) ?? this.getNewestPredicate();
+    if (!predicate) throw new Error('No predicate found');
+
+    this.predicateAddress = predicate.getRoot();
+    this.predicateAccount = predicate;
+
+    return this.predicateAccount;
+  }
+
+  protected subscribe(listener: () => void) {
+    this.subscriptions.push(listener);
+  }
+
+  protected async prepareTransaction(
+    address: string,
+    transaction: TransactionRequestLike,
+  ): Promise<PreparedTransaction> {
+    if (!(await this.isConnected())) {
+      throw Error('No connected accounts');
+    }
+
+    if (!this.predicateAccount) {
+      throw Error('No predicate account found');
+    }
+
+    const { fuelProvider } = await this.getProviders();
+    const chainId = fuelProvider.getChainId();
+    const walletAccount = this.predicateAccount.getAccountAddress(
+      address,
+      await this.walletAccounts(),
+    );
+    if (!walletAccount) {
+      throw Error(`No account found for ${address}`);
+    }
+
+    const transactionRequest = transactionRequestify(transaction);
+    const newestPredicate = this.getNewestPredicate();
+    if (!!newestPredicate && !this.predicateAccount.equals(newestPredicate)) {
+      const predicateAddress =
+        newestPredicate.getPredicateAddress(walletAccount);
+
+      if (transactionRequest.getChangeOutputs().length > 0) {
+        transactionRequest.outputs.forEach((output) => {
+          if (
+            output.type !== OutputType.Change ||
+            !this.isAddressPredicate(output.to)
+          )
+            return;
+          output.to = Address.fromAddressOrString(predicateAddress).toB256();
```

Comment:
this makes us mentally need to invert the conditions to understand what's going on, better to do smth like:

```
if (
            output.type === OutputType.Change &&
            this.isAddressPredicate(output.to)
          ) {
            output.to = Address.fromAddressOrString(predicateAddress).toB256()
          }
```